File: src/ZmqClient.cpp
#include "ZmqClient.h"
#include <iostream>

namespace {
std::mutex _mutex;
std::string createAddr(const std::string& ip, uint16_t port)
{
    return "tcp://" + ip + ':' + std::to_string(port);
}
}

bool ZmqClient::setAddr(const std::string& ip, uint16_t port, int rcvTimeout)
{
    if (ip.empty()) return false;

    if (!_addr.empty())
    {
        _socket.disconnect(_addr);
    }
    _socket.setsockopt(ZMQ_RCVTIMEO, rcvTimeout);
    _socket.setsockopt(ZMQ_SNDTIMEO, 1000);
    _socket.setsockopt(ZMQ_LINGER, 0);
    _addr = createAddr(ip, port);
    std::cout << "Connect to " << _addr.c_str() << std::endl;
    _socket.connect(_addr);
    return true;
}

void ZmqClient::disconnect()
{
    _socket.disconnect(_addr);
    _addr.clear();
}

zmq::message_t ZmqClient::sendMsg()
{
    zmq::message_t message(_reqBuf.data(), _reqBuf.size());
    try
    {
        if (!_socket.send(message))
        {
            std::cout << "Send failed!" << _addr.c_str() << std::endl;
        }
        else if (!_socket.recv(&message))
        {
            std::cout << "Recv failed!" << _addr.c_str() << std::endl;
        }
        else
        {
            return message;
        }
    } catch (const zmq::error_t& t)
    {
        std::cout << _addr.c_str() << "socket error: " << t.what() << std::endl;
    }
    return {};
}

File: src/ZmqClient.py
#!/usr/bin/env python
import zmq
import sys

def createAddr(ip, port):
    return "tcp://" + ip + ":" + str(port)


class ZmqClient:
    def __init__(self):
        self.__addr = ""
        self.__reqBuf = ""
        self.__context = zmq.Context()
        self.__socket = self.__context.socket(zmq.REQ)

    def setAddr(self, ip, port, rcvTimeout):
        if len(ip) == 0:
            return False
        if len(self.__addr) != 0:
            self.__socket.disconnect()
        self.__socket.setsockopt(zmq.RCVTIMEO, rcvTimeout)
        self.__socket.setsockopt(zmq.SNDTIMEO, 1000)
        self.__socket.setsockopt(zmq.LINGER, 0)
        self.__addr = createAddr(ip, port)
        print("connect to %s" % (self.__addr))
        self.__socket.connect(self.__addr)
        return True

    def disconnect(self):
        self.__socket.disconnect(self.__addr)
        self.__addr = ""

    def empty(self):
        return len(self.__addr) == 0

    def sendReq(self, request):
        try:
            self.__reqBuf = request
            if len(self.__addr) == 0:
                return {}
            reply = self.__sendMsg()
            if len(reply) == 0:
                return {}
            return reply
        except Exception as e:
            print("Network Error! Please check your ip address and connection!")
            sys.exit(0)

    def __sendMsg(self):
        try:
            message = self.__reqBuf
            self.__socket.send_string(message)
            return self.__socket.recv()
        except Exception as e:
            print("Network Error! Please check your ip address and connection!")
            sys.exit(0)
